File: HackerRank/average-modulo.cpp
#include <bits/stdc++.h>

using namespace std;

int gcd(int a, int b) {
  if (b == 0)
    return a;
  else
    return gcd(b, a%b);
}


struct Rat{
  int num;
  int den;

  Rat() : num{0}, den{1} {}
  Rat(int num, int den) : num{num}, den{den} {
    reduce();
  }

  Rat operator+(const Rat& other) {
    return Rat(num * other.den + den * other.num, den * other.den);
  }

  void reduce() {
    int this_gcd = gcd(num, den);
    num /= this_gcd;
    den /= this_gcd;
  }
};


int T;
int N, p, k;
int arr[50001];
Rat prefix[50001];

bool is_attainable(Rat avg)

int main() {
  scanf("%d", &T);
  for (int t = 0; t < T; ++t) {
    scanf("%d%d%d", &N, &p, &k);
    for (int i = 0; i < N; ++i) {
      scanf("%d", arr+i);
    }
  }
  
  return 0;
}

File: HackerRank/average-modulo.py
class Rat:
    def __init__(self, num: int, den: int):
        self.num = num
        self.den = den
        self.reduce()

    def __add__(self, other: Rat):
        return Rat(self.num * other.den + other.num * self.den,
                   self.den * other.den)
    def __radd__(self, other):
        self.num = self.num * other.den + other.num * self.den
        self.den = self.den * other.den
        self.reduce()

    def __sub__(self, other: Rat):
        return Rat(self.num * other.den - other.num * self.den,
                   self.den * other.den)
    def __rsub__(self, other):
        self.num = self.num * other.den - other.num * self.den
        self.den = self.den * other.den
        self.reduce()

    def __mul__(self, other: Rat):
        return Rat(self.num * other.num,
                   self.den * other.den)
    def __rmul__(self, other):
        self.num = self.num * other.num
        self.den = self.den * other.den
        self.reduce()

    def __truediv__(self, other: Rat):
        return Rat(self.num * other.den,
                   self.den * other.num)
    def __rtruediv__(self, other):
        self.num = self.num * other.den
        self.den = self.den * other.num
        self.reduce()

    def __eq__(self, other):
        return self.num == other.num and self.den == other.den
    def __gt__(self, other):
        return self.num*other.den < other.num*self.den
    def __ge__(self, other):
        return self > other or self == other
    def __lt__(self, other):
        return not (self >= other)
    def __le__(self, other):
        return not (self > other)

    @staticmethod
    def gcd(a: int, b: int):
        if b == 0:
            return a
        else:
            return Rat.gcd(b, a % b)

    def reduce(self):
        n = -self.num if self.num < 0 else self.num
        d = self.den

        gcd = Rat.gcd(n, d)
        self.num /= gcd
        self.den /= gcd

T = int(input())
for _ in range(T):
  N, p, k = list(map(int, input().split()))
  a = list(map(int, input().split()))
  

